src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -80,18 +78,199 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
+                        tokenList.Add(tokens.Data[tokenIndex]);
+                    }
+
+                    tokenList.Add(tokens.Data[tokenIndex + 1]);
+                    tokenList.Add(tokens.Data[tokenIndex + 2]);
+                    tokenList.Add(tokens.Data[tokenIndex + 3]);
+                    tokenList.Add(tokens.Data[tokenIndex + 4]);
+                }
+            }
+
+            // Add the cached range tokens to the results.
+            for (var cachedTokenIndex = 0; cachedTokenIndex < cachedTokens.Count; cachedTokenIndex++)
+            {
+                if (cachedTokenIndex == 0)
+                {
+                    tokenList.Add(cachedTokens[0] - absoluteLine);
+                    absoluteLine = cachedTokens[0];
+                }
+                else
+                {
+                    tokenList.Add(cachedTokens[cachedTokenIndex]);
+
+                    if (cachedTokenIndex % 5 == 0)
+                    {
+                        absoluteLine += cachedTokens[cachedTokenIndex];
+                    }
+                }
+            }
+
+            // Compute the ending range tokens (if applicable).
+            if (range.End.Line != cachedRange.End.Line)
+            {
+                var partialRange = new Range { Start = cachedRange.End, End = range.End };
+                var tokens = await ComputeAndCacheTokensAsync(
+                    textDocumentIdentifier, documentSnapshot, documentVersion, semanticVersion,
+                    partialRange, cancellationToken).ConfigureAwait(false);
+
+                var firstToken = true;
+                var endAbsoluteLine = 0;
+                for (var tokenIndex = 0; tokenIndex < tokens.Data.Length; tokenIndex += 5)
+                {
+                    if (tokenIndex == 0)
+                    {
+                        endAbsoluteLine = tokens.Data[0];
+                    }
+                    else
+                    {
+                        endAbsoluteLine += tokens.Data[tokenIndex];
+                    }
+
+                    // Skip tokens that are out of the range
+                    if (endAbsoluteLine < cachedRange.End.Line || endAbsoluteLine >= range.End.Line)
+                    {
+                        continue;
+                    }
+
+                    if (firstToken)
+                    {
+                        tokenList.Add(endAbsoluteLine - absoluteLine);
+                        firstToken = false;
+                    }
+                    else
+                    {
+                        tokenList.Add(tokens.Data[tokenIndex]);
+                    }
+
+                    tokenList.Add(tokens.Data[tokenIndex + 1]);
+                    tokenList.Add(tokens.Data[tokenIndex + 2]);
+                    tokenList.Add(tokens.Data[tokenIndex + 3]);
+                    tokenList.Add(tokens.Data[tokenIndex + 4]);
+                }
+            }
+
+            return new SemanticTokens { Data = tokenList.ToImmutableArray() };
```

Comment:
basically, i really have no idea what's goign on here.

Reply:
Adding some comments to my next push which should hopefully help clarify things